x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/ChainingInputStreamTests.java

```diff
@@ -0,0 +1,573 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.repositories.encrypted;
+
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.test.ESTestCase;
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+import org.mockito.Mockito;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class ChainingInputStreamTests extends ESTestCase {
+
+    public void testSkipAcrossComponents() throws Exception {
+        ChainingInputStream test = new ChainingInputStream() {
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                return null;
+            }
+        };
+        byte[] b1 = new byte[1 + Randomness.get().nextInt(16)];
+        Randomness.get().nextBytes(b1);
+        test.currentIn = new ByteArrayInputStream(b1);
+        long nSkip = test.skip(b1.length + 1 + Randomness.get().nextInt(16));
+        assertThat(nSkip, Matchers.is((long)b1.length));
+        byte[] b2 = new byte[1 + Randomness.get().nextInt(16)];
+        Randomness.get().nextBytes(b2);
+        test = new ChainingInputStream() {
+            boolean second = false;
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                if (false == second) {
+                    second = true;
+                    return new ByteArrayInputStream(b2);
+                } else {
+                    return null;
+                }
+            }
+        };
+        test.currentIn = new ByteArrayInputStream(b1);
+        long skipArg = b1.length + 1 + Randomness.get().nextInt(b2.length);
+        nSkip = test.skip(skipArg);
+        assertThat(nSkip, Matchers.is(skipArg));
+        byte[] rest = test.readAllBytes();
+        assertThat((long)rest.length, Matchers.is(b1.length + b2.length - nSkip));
+        for (int i = rest.length - 1; i >= 0; i--) {
+            assertThat(rest[i], Matchers.is(b2[i + (int)nSkip - b1.length]));
+        }
+    }
+
+    public void testEmptyChain() throws Exception {
```

Comment:
Can we have the same set of tests for empty component streams too like:
```
final Iterator<InputStream> iterator = Arrays.asList(
    randomArray(1, 5, InputStream[]::new, InputStream::nullInputStream)).iterator();
ChainingInputStream zeroByteStream = new ChainingInputStream() {
    InputStream nextComponent(InputStream currentComponentIn) throws IOException {
        if (iterator.hasNext()) {
            return iterator.next();
        } else {
            return null;
        }
     } 
};
```
`testEmptyInputStreamComponents` doesn't seem to cover the case of _everything_ being empty.

Reply:
Agreed, thanks.